File: scala/grade-school/src/main/scala/School.scala
import scala.collection.immutable.TreeMap

class School {

  type Grade = Int
  type Name = String
  type Db = TreeMap[Grade, List[Name]]

  private val emptyNames = List.empty[Name]
  private val empty = TreeMap[Grade, List[Name]]()

  var db: Db = empty

  def add(name: Name, grade: Grade): Unit = {
    val updatedNames = db.getOrElse(grade, emptyNames) :+ name
    db = db + (grade -> updatedNames)
  }

  def grade(grade: Grade): List[Name] = db.getOrElse(grade, emptyNames)

  def sorted: Db = db.foldLeft(empty)(sortNames)

  def sortNames(map: Db, entry: (Grade, List[Name])): Db = {
    val (k, names) = entry
    map + (k -> names.sorted)
  }
}

File: haskell/grade-school/School.hs
module School (School, empty, grade, sorted, add) where

import qualified Data.Map as M
import Data.List(sort)

data School = School { impl :: M.Map Int [String] }

add :: Int -> String -> School -> School
add grade name school = School $ M.insert grade newValue imp
  where
    imp = impl school
    newValue = name : M.findWithDefault [] grade imp


empty :: School
empty = School $ M.fromList []

sorted :: School -> [(Int, [String])]
sorted = M.toAscList . M.map sort . impl

grade :: Int -> School -> [String]
grade n school = M.findWithDefault [] n (impl school)